src/ida_lunde_naalsund_ex/ex06/logistic_regression.py

```diff
@@ -0,0 +1,270 @@
+# -*- coding: utf-8 -*-
+
+__author__ = "Ida Lunde Naalsund, Yngve Mardal Moe"
+__email__ = "idna@nmbu.no, yngve.m.moe@gmail.com"
+
+import numpy as np
+from sklearn.base import BaseEstimator, ClassifierMixin
+from sklearn.exceptions import NotFittedError
+from sklearn.utils import check_random_state, check_X_y
+
+
+def sigmoid(z):
+    """Perform a logistic transform on the input.
+    This function applies the sigmoidal function element-wise to all
+    elements of `z`.
+    Parameters
+    ----------
+    z : np.ndarray
+        Logit to transform.
+    Returns
+    -------
+    sigmoidal_transformed_z : np.ndarray
+        Transformed input.
+    """
+
+    return 1 / (1 + np.exp(-z))
+
+
+def predict_proba(coef, X):
+    r"""Predict the class probabilities for each data point in :math:`X`.
+    Estimate which class each data point in X corresponds to. This is done
+    according to the following formula.
+    .. math::
+        \hat{y}_i = \sigma(\mathbf{x}_i^T \mathbf{w}),
+    where :math:`x_i` is the i-th row in :math:`X` and :math:`\sigma` is
+    the sigmoidal function. Alternatively, in matrix-vector form:
+    .. math::
+        \hat{\mathbf{y}} = \sigma(X \mathbf{w}).
+    Parameters
+    ----------
+    coef : np.ndarray(shape=(r,))
+        The weight vector, :math:`w`
+    X : np.ndarray(shape=(n, r))
+        The data matrix (aka design or measurement matrix)
+    Returns
+    -------
+    p : np.ndarray(shape(n,))
+        The predicted class probabilities.
+    """
+
+    return sigmoid(np.dot(X, coef))
+
+
+def logistic_gradient(coef, X, y):
+    """Returns the gradient of a logistic regression model.
+
+    Parameters
+    ----------
+    coef : np.ndarray(shape=(r,))
+        The weight vector, :math:`w`
+    X : np.ndarray(shape=(n, r))
+        The data matrix (aka design or measurement matrix)
+    y : np.ndarray(shape=(n,))
+        The true class labels for each data point.
+    Returns
+    -------
+    gradient : np.ndarray(shape=(r,))
+        The gradient of the cross entropy loss related to the linear
+        logistic regression model.
+    """
+
+    return np.dot(X.T, (y - predict_proba(coef, X)))
```

Comment:
If you experienced a low accuracy, its because this is supposed to be `predicted - y`, not the opposite. If you `-=` coeff, the model will diverge, rather than converge. You can check this, by changing it and checking the accuracy by running `__main__`

Reply:
This is from the updated task description, btw.